### PR TITLE
sql

### PR DESCRIPTION
**Pull request by jgaleotti**
Regex genes for SIMILAR_TO and LIKE check constraints

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/database/DbActionGeneBuilder.kt [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+                .map { createGeneForPostgresLike(it) }
+                .map { it.disjunctions }
+                .map { it.disjunctions }
+                .flatten()
+        val disjunctionListRxGene = DisjunctionListRxGene(disjunctions = disjunctionRxGenes)
+        return RegexGene(columnName, disjunctions = disjunctionListRxGene)
+    }
+
+    fun buildSimilarToRegexGene(columnName: String, similarToPatterns: List<String>): RegexGene {
+        val joinedPattern = similarToPatterns.map { c -> "(%s)".format(c) }.joinToString("|")
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/database/DbActionGeneBuilder.kt]: i do not understand this `"(%s)".format(c)`. can u add a comment?

File: core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt [hunk 2]
```diff
@@ -86,7 +90,9 @@ class SqlInsertBuilder(
                         unique = c.unique,
                         lowerBound = lowerBound,
                         upperBound = upperBound,
-                        enumValuesAsStrings = enumValuesAsStrings
+                        enumValuesAsStrings = enumValuesAsStrings,
+                        similarToPatterns = similarToPatterns,
+                        likePatterns = likePatterns
                 )
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt]: i have seen LIKE and SIMILAR TO (plus other kinds of regexs) to be different among databases... :(
i guess that somehow we need to store info of which database we are using

File: core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt [hunk 3]
```diff
@@ -154,6 +160,84 @@ class SqlInsertBuilder(
 
     }
 
+    private class ConstraintCollector : TableConstraintVisitor<List<TableConstraint>, Void> {
+
+        override fun visit(constraint: IsNotNullConstraint?, argument: Void?): List<TableConstraint> {
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt]: why in these visitors are you passing nullable inputs with `?` ? even if extending a Java class, putting them non-nullable should still compile

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+
+    override fun randomize(randomness: Randomness, forceNewValue: Boolean, allGenes: List<Gene>) {
+        objectGene.randomize(randomness, forceNewValue, allGenes)
+    }
+
+
+    override fun getValueAsPrintableString(previousGenes: List<Gene>, mode: String?, targetFormat: OutputFormat?): String {
+        return "\"%s\"".format(objectGene.getValueAsPrintableString(previousGenes, mode, targetFormat))
+    }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt]: can u comment this `"\"%s\""`?

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            objectGene = this.objectGene.copy() as ObjectGene)
+
+
+    override fun randomize(randomness: Randomness, forceNewValue: Boolean, allGenes: List<Gene>) {
+        objectGene.randomize(randomness, forceNewValue, allGenes)
+    }
+
+
+    override fun getValueAsPrintableString(previousGenes: List<Gene>, mode: String?, targetFormat: OutputFormat?): String {
+        return "\"%s\"".format(objectGene.getValueAsPrintableString(previousGenes, mode, targetFormat))
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt]: shouldn't the `mode` here always being JSON?

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlPrimaryKeyGene.kt [hunk 6]
```diff
@@ -55,8 +55,9 @@ class SqlPrimaryKeyGene(name: String,
 
     override fun getVariableName() = gene.getVariableName()
 
-    override fun flatView(excludePredicate: (Gene)-> Boolean): List<Gene> {
-        return if(excludePredicate(this)) listOf(this) else listOf(this).plus(gene.flatView(excludePredicate))
+    override fun flatView(excludePredicate: (Gene) -> Boolean): List<Gene> {
+        return if (excludePredicate(this)) listOf() else
+            listOf(this).plus(gene.flatView(excludePredicate))
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/SqlPrimaryKeyGene.kt]: isn't this a regression bug where ` listOf(this)` is wrongly replaced with ` listOf()`?

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlUUIDGene.kt [hunk 7]
```diff
@@ -27,19 +32,21 @@ class UUIDGene(
     }
 
     override fun getValueAsRawString(): String {
-        return UUID(mostSigBits.value, leastSigBits.value).toString()
+        return "{%s}".format(getValueAsUUID())
     }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/SqlUUIDGene.kt]: please comment on `"{%s}".format`

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/RegexGene.kt [hunk 8]
```diff
@@ -21,10 +20,23 @@ class RegexGene(
...
+        val rawValue = getValueAsRawString()
+        when {
+            (targetFormat == null) -> return "\"$rawValue\""
+            targetFormat.isKotlin() -> return "\"$rawValue\""
+                    .replace("\\", "\\\\")
+                    .replace("$", "\\$")
+            else -> return "\"$rawValue\""
+                    .replace("\\", "\\\\")
+        }
     }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/regex/RegexGene.kt]: add a TODO comment: likely these issues with Kotlin (and other languages) escapes would apply to all printed string, and could be moved into a single place

File: core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt (hunk 3 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt]: You are right. I removed the ? modifier on all the constraints

File: core/src/main/kotlin/org/evomaster/core/database/DbActionGeneBuilder.kt (hunk 1 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/database/DbActionGeneBuilder.kt]: I changed the implementation of buildSimilarToRegexGene() to match buildLikeRegexGene()

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlPrimaryKeyGene.kt (hunk 6 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/SqlPrimaryKeyGene.kt]: You're right, sorry for this.

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt (hunk 5 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt]: You are right. In fact, that's what happens with SqlXMLGene (mode is always "xml")

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/RegexGene.kt (hunk 8 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/regex/RegexGene.kt]: Added a //TODO comment for future refactor with StringGene.getValueAsPrintableString()

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlUUIDGene.kt (hunk 7 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/SqlUUIDGene.kt]: I checked that postgresql allows inserting uuid values without any enclosing braces {...}

File: core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt (hunk 4 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/SqlJSONGene.kt]: I cloned the same code (with a proper TODO comment for future refactor)  than StringGene.getValueAsPrintableString()

File: core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt (hunk 2 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt]: I am adding databaseType to the column, so it flows to the RegexHandler.

**Comment by jgaleotti** [comment]
would you like to check the changes I made to the PR?

**Comment by arcuri82** [comment]
yes, thanks. i ll try to check them by tomorrow's meeting

File: core/src/main/kotlin/org/evomaster/core/parser/RegexHandler.kt [hunk 9]
```diff
@@ -44,7 +45,7 @@ object RegexHandler {
      * This would throw an exception if regex is invalid, or if it
      * has features we do not support yet
      */
-    fun createGeneForPostgresLike(regex: String): RegexGene{
+    fun createGeneForPostgresLike(regex: String, databaseType: DatabaseType? = null): RegexGene {
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/parser/RegexHandler.kt]: this is not what I meant :)
notice that the method `createGeneForPostgresLike` as `Postgres` inside its name.
so, if `databaseType` is Postgres, then we call this method.
Note that, for the moment, we only support Postgres.

File: core/src/main/kotlin/org/evomaster/core/parser/RegexHandler.kt [hunk 10]
```diff
@@ -66,7 +67,7 @@ object RegexHandler {
      * This would throw an exception if regex is invalid, or if it
      * has features we do not support yet
      */
-    fun createGeneForPostgresSimilarTo(regex: String): RegexGene{
+    fun createGeneForPostgresSimilarTo(regex: String, databaseType: DatabaseType? = null): RegexGene {
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/parser/RegexHandler.kt]: see comment above

File: core/src/main/kotlin/org/evomaster/core/search/gene/StringGene.kt [hunk 11]
```diff
@@ -32,13 +33,18 @@ class StringGene(
...
-                                                            .replace("\\", "\\\\")
-                                                            .replace("$", "\\$")
-            else -> return "\"$value\""
-                    .replace("\\", "\\\\")
+        val rawValue = getValueAsRawString()
+        if (mode != null && mode.equals("xml")) {
+            return StringEscapeUtils.escapeXml(rawValue)
+        } else {
+            when {
+                (targetFormat == null) -> return "\"$rawValue\""
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/StringGene.kt]: add a  `// TODO: refactor`
at later stage, we will need a general way to handle with this issue

File: core/src/main/kotlin/org/evomaster/core/parser/RegexHandler.kt (hunk 9 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/parser/RegexHandler.kt]: Sorry, I missed the "Postgres" in the method name. I will fix that.

File: core/src/main/kotlin/org/evomaster/core/search/gene/StringGene.kt (hunk 11 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/StringGene.kt]: ok

**Comment by arcuri82** [comment]
let me know once you are done with your changes, and I will review them

**Comment by jgaleotti** [comment]
Done! I pushed the merged files into the sql branch

**Pull request merged**